src/libutil/error.cc

```diff
@@ -50,6 +50,29 @@ std::ostream & operator <<(std::ostream & str, const AbstractPos & pos)
     return str;
 }
 
+inline bool operator<(const AbstractPos& lhs, const AbstractPos& rhs)
+{
+    return lhs.line < rhs.line
+        || lhs.column < rhs.column;
+}
+inline bool operator> (const AbstractPos& lhs, const AbstractPos& rhs) { return rhs < lhs; }
+inline bool operator<=(const AbstractPos& lhs, const AbstractPos& rhs) { return !(lhs > rhs); }
+inline bool operator>=(const AbstractPos& lhs, const AbstractPos& rhs) { return !(lhs < rhs); }
+
+inline bool operator<(const Trace& lhs, const Trace& rhs)
+{
+    if (!lhs.pos && rhs.pos) {
+        return true;
+    }
+
+    return (lhs.pos && rhs.pos && *lhs.pos < *rhs.pos)
+        || lhs.hint.str() < rhs.hint.str()
+        || lhs.frame < rhs.frame;
+}
+inline bool operator> (const Trace& lhs, const Trace& rhs) { return rhs < lhs; }
+inline bool operator<=(const Trace& lhs, const Trace& rhs) { return !(lhs > rhs); }
+inline bool operator>=(const Trace& lhs, const Trace& rhs) { return !(lhs < rhs); }
```

Comment:
These should probably be in an anonymous namespace as they're not forward-declared nor used from any other source file (c++ people invent a clang-tidy rule for this challenge).

Reply:
I get an error if I try that. I think I have to actually make them public in the same namespace as `Trace`/`AbstractPos` to be visible?

https://stackoverflow.com/questions/55390265/defining-a-custom-equality-operator-in-an-anonymous-namespace
https://stackoverflow.com/questions/52629859/adl-for-class-in-anonymous-namespace

Reply:
Interesting; the following program compiles with `g++`:

```cpp
namespace meow {
class Blah {};
}

namespace {
    inline bool operator>(const meow::Blah& a, const meow::Blah& b) { return true; }
};

int main(void) {
    meow::Blah{} > meow::Blah{};
}
```

I don't know what's going on, but the code is not problematic as-is; it could also possibly be moved to the header where Trace is defined.

Reply:
Done.